keylime/src/crypto/x509.rs

```diff
@@ -0,0 +1,592 @@
+    }
+
+    /// Set the hash algorithm to be used when signing the certificate
+    ///
+    /// # Arguments:
+    ///
+    /// * hash_algorithm (MessageDigest): The hash algorithm to be used when signing the certificate
+    pub fn hash_algorithm(
+        &'a mut self,
+        hash_algorithm: MessageDigest,
+    ) -> &mut CertificateBuilder<'a> {
+        self.hash_algorithm = Some(hash_algorithm);
+        self
+    }
+
+    /// Set the certificate start of validity, in days from now
+    ///
+    /// # Arguments:
+    ///
+    /// * days_from_now (u32): The number of days from now when the built certificate will become
+    /// valid
+    pub fn not_before(
+        &'a mut self,
+        days_from_now: u32,
+    ) -> &mut CertificateBuilder<'a> {
+        self.not_before = Some(days_from_now);
+        self
+    }
+
+    /// Set the certificate expiration date, in days from now
+    ///
+    /// # Arguments:
+    ///
+    /// * days_from_now (u32): The number of days from now when the built certificate will expire
+    pub fn not_after(
+        &'a mut self,
+        days_from_now: u32,
+    ) -> &mut CertificateBuilder<'a> {
+        self.not_after = Some(days_from_now);
+        self
+    }
+
+    /// Set the certificate X.509 standard version.
+    ///
+    /// # Arguments:
+    ///
+    /// * version (i32): The version number. Note that the version is zero-indexed, meaning passing
+    /// the value `2` corresponds to the version 3 of the X.509 standard
+    ///
+    /// If not called, the version 3 of the X.509 standard will be used
+    pub fn version(
+        &'a mut self,
+        version: i32,
+    ) -> &mut CertificateBuilder<'a> {
+        self.version = Some(version);
+        self
+    }
+
+    /// Set the private key associated with the certificate
+    ///
+    /// # Arguments:
+    ///
+    /// * private_key (PKey<Private>): The private key to be associated with the certificate
+    pub fn private_key(
+        &'a mut self,
+        private_key: &'a PKey<Private>,
+    ) -> &mut CertificateBuilder<'a> {
+        self.private_key = Some(private_key);
+        self
+    }
+
+    /// Set DNS names to add to the Subject Alternative Name
+    ///
+    /// # Arguments:
+    ///
+    /// * dns_names (Vec<String>): A Vec<String> containing DNS names to add to the certificate Subject
+    /// Alternative Name
+    pub fn add_dns_names(
+        &'a mut self,
+        dns_names: Vec<String>,
+    ) -> &mut CertificateBuilder<'a> {
+        match &mut self.dns_names {
+            None => {
+                self.dns_names = Some(dns_names);
+            }
+            Some(v) => {
+                for name in dns_names {
+                    v.push(name);
+                }
+            }
+        }
+        self
+    }
+
+    /// Set additional IPs to add to the Subject Alternative Name
+    ///
+    /// # Arguments:
+    ///
+    /// * ips: (Vec<String>): A Vec<String> containing IPs to add to the certificate Subject
```

Comment:
Does this need to be owned? If this is `&[&str]`, the callers don't need to create owned values by themselves: `.add_ips(&vec!["192.168.0.1"])` instead of `.add_ips(vec!["192.168.0.1".to_string()])`.